core/Session/SessionAuth.php

```diff
@@ -0,0 +1,166 @@
+        $userModel = $this->userModel;
+
+        $userForSession = $sessionFingerprint->getUser();
+        if (empty($userForSession)) {
+            return $this->makeAuthFailure();
+        }
+
+        $user = $userModel->getUser($userForSession);
+        if (empty($user)
+            || $user['login'] !== $userForSession // sanity check in case there's a bug in getUser()
+        ) {
+            return $this->makeAuthFailure();
+        }
+
+        if (!$sessionFingerprint->isMatchingCurrentRequest()) {
+            $this->initNewBlankSession($sessionFingerprint);
+            return $this->makeAuthFailure();
+        }
+
+        $tsPasswordModified = !empty($user['ts_password_modified']) ? $user['ts_password_modified'] : null;
+        if ($this->isSessionStartedBeforePasswordChange($sessionFingerprint, $tsPasswordModified)) {
+            $this->destroyCurrentSession($sessionFingerprint);
+            return $this->makeAuthFailure();
+        }
+
+        return $this->makeAuthSuccess($user);
+    }
+
+    private function isSessionStartedBeforePasswordChange(SessionFingerprint $sessionFingerprint, $tsPasswordModified)
+    {
+        // sanity check, make sure users can still login if the ts_password_modified column does not exist
+        if ($tsPasswordModified === null) {
+            return false;
+        }
+
+        // if the session start time doesn't exist for some reason, log the user out
+        $sessionStartTime = $sessionFingerprint->getSessionStartTime();
+        if (empty($sessionStartTime)) {
+            return true;
+        }
+
+        return $sessionStartTime < Date::factory($tsPasswordModified)->getTimestampUTC();
+    }
+
+    private function makeAuthFailure()
+    {
+        return new AuthResult(AuthResult::FAILURE, null, null);
+    }
+
+    private function makeAuthSuccess($user)
+    {
+        $this->setTokenAuth($user['token_auth']);
+
+        $isSuperUser = (int) $user['superuser_access'];
+        $code = $isSuperUser ? AuthResult::SUCCESS_SUPERUSER_AUTH_CODE : AuthResult::SUCCESS;
+
+        return new AuthResult($code, $user['login'], $user['token_auth']);
+    }
+
+    private function initNewBlankSession(SessionFingerprint $sessionFingerprint)
+    {
+        // this user should be using a different session, so generate a new ID
+        // NOTE: Zend_Session cannot be used since it will destroy the old
+        // session.
+        if ($this->shouldDestroySession) {
+            session_regenerate_id();
+        }
+
+        // regenerating the ID will create a new session w/ a new ID, but will
+        // copy over the existing session data. we want the new session for the
+        // unauthorized user to be different, so we clear the session fingerprint.
+        $sessionFingerprint->clear();
+    }
+
+    private function destroyCurrentSession(SessionFingerprint $sessionFingerprint)
+    {
+        // Note: can't use Session::destroy() since Zend prohibits starting a new session
```

Comment:
just wondering: why are we starting a new session after it was destroyed? I think this was not needed before this PR. Ideally destroying session as early as possible be great as it can increase performance quite a bit. (haven't really looked too deep into this, just wondering)

Reply:
I'm not exactly sure why anymore, but Zend kept throwing an exception when I called Session::destroy(). It might not be necessary, but I'll have to switch it back and check to figure out where the second session start happened. (This is what you're asking about, right?)

Also note (and you might be aware of this, just mentioning in case it wasn't clear), the session is destroyed in the same place as it was before in FrontController. All of this session stuff happens before that, so it shouldn't have an effect on when it gets closed.

Reply:
Did a check and when I use `Session::destroy()`, this code results in an exception being thrown. The message I see is: `The session was explicitly destroyed during this request, attempting to re-start is not allowed.`

And the stacktrace is:

```
#0 /usr/src/piwik/libs/Zend/Session/Namespace.php(143): Zend_Session::start(true)
#1 /usr/src/piwik/core/Session/SessionNamespace.php(34): Zend_Session_Namespace->__construct('Login.login', false)
#2 /usr/src/piwik/core/Nonce.php(39): Piwik\Session\SessionNamespace->__construct('Login.login')
#3 /usr/src/piwik/plugins/Login/Controller.php(133): Piwik\Nonce::getNonce('Login.login')
#4 /usr/src/piwik/plugins/Login/Controller.php(115): Piwik\Plugins\Login\Controller->configureView(Object(Piwik\View))
#5 [internal function]: Piwik\Plugins\Login\Controller->login('You can't acces...')
#6 /usr/src/piwik/core/FrontController.php(517): call_user_func_array(Array, Array)
#7 /usr/src/piwik/core/FrontController.php(140): Piwik\FrontController->doDispatch('Login', 'login', Array)
#8 /usr/src/piwik/plugins/Login/Login.php(63): Piwik\FrontController->dispatch('Login', 'login', Array)
#9 [internal function]: Piwik\Plugins\Login\Login->noAccess(Object(Piwik\NoAccessException))
#10 /usr/src/piwik/core/EventDispatcher.php(141): call_user_func_array(Array, Array)
#11 /usr/src/piwik/core/Piwik.php(701): Piwik\EventDispatcher->postEvent('User.isNotAutho...', Array, true, NULL)
#12 /usr/src/piwik/core/FrontController.php(153): Piwik\Piwik::postEvent('User.isNotAutho...', Array, true)
#13 /usr/src/piwik/core/dispatch.php(34): Piwik\FrontController->dispatch()
#14 /usr/src/piwik/index.php(27): require_once('/usr/src/piwik/...')
#15 {main}
```

Looks like it's because of the nonce that's used on the Login form. `Zend_Session` will throw if the session was destroyed through Zend: https://github.com/piwik/piwik/blob/3.x-dev/libs/Zend/Session.php#L420

I'm not sure if there's a better way to do this than to regenerate the ID, but I can't think of a (quick) better idea right now.

Reply:
I'll update the comment w/ regard to the nonce issue.

Reply:
> This is what you're asking about, right?

Yes I was basically wondering why this was not a problem before, and why it is now as there seems to be a difference (unless we never noticed the problem before or so)

Reply:
(missed your last comment) Before I don't think we were ever destroying the session (we also weren't really using it since it could disappear at any time). It's only a problem now that I wanted to destroy invalid sessions.